src/libcore/num/flt2dec/strategy/grisu.rs

```diff
@@ -0,0 +1,749 @@
+// Copyright 2015 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+/*!
+Rust adaptation of Grisu3 algorithm described in [1]. It uses about
+1KB of precomputed table, and in turn, it's very quick for most inputs.
+
+[1] Florian Loitsch. 2010. Printing floating-point numbers quickly and
+    accurately with integers. SIGPLAN Not. 45, 6 (June 2010), 233-243.
+*/
+
+use prelude::*;
+use num::{Int, Float};
+
+use num::flt2dec::{Decoded, MAX_SIG_DIGITS, round_up};
+
+/// A custom 64-bit floating point type, representing `f * 2^e`.
+#[derive(Copy, Clone, Debug)]
+#[doc(hidden)]
+pub struct Fp {
+    /// The integer mantissa.
+    pub f: u64,
+    /// The exponent in base 2.
+    pub e: i16,
```

Comment:
I haven't read the Grisu paper carefully, but in its discussion of the `diy_fp` struct, it notes that (1.) the paper's definition uses an unlimited range value for `e_x` to simplify the proofs, (2.) in practice the exponent type must only have "slightly greater range than the input exponent" and (3.) for IEEE doubles which have 11-bit exponents, a *32-bit* signed integer is by far big enough.

The combination of 2. and 3. leads me to wonder whether we could get into trouble using only `i16` here.  I know you exhaustively tested your algorithm for `f32`, but how confident are you that `i16` suffices for `f64` ?